server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java

```diff
@@ -776,6 +789,35 @@ public ClusterState execute(BatchExecutionContext<StartedShardUpdateTask> batchE
             return maybeUpdatedState;
         }
 
+        private ClusterState maybeRemoveIndexRefreshBlocks(
+            ClusterState clusterState,
+            @Nullable Set<Index> indicesWithUnpromotableShardsStarted
+        ) {
+            if (indicesWithUnpromotableShardsStarted == null) {
+                return clusterState;
+            }
+
+            ClusterBlocks.Builder clusterBlocksBuilder = null;
+            for (Index indexWithUnpromotableShardsStarted : indicesWithUnpromotableShardsStarted) {
+                String indexName = indexWithUnpromotableShardsStarted.getName();
+                assert clusterState.blocks().hasIndexBlock(indexName, INDEX_REFRESH_BLOCK);
+
+                var indexRoutingTable = clusterState.routingTable().index(indexWithUnpromotableShardsStarted);
+                if (indexRoutingTable.readyForSearch(clusterState)) {
+                    if (clusterBlocksBuilder == null) {
+                        clusterBlocksBuilder = ClusterBlocks.builder(clusterState.blocks());
+                    }
+                    clusterBlocksBuilder.removeIndexBlock(indexName, INDEX_REFRESH_BLOCK);
+                }
+            }
+
+            if (clusterBlocksBuilder == null) {
```

Comment:
Is that possible? If `indicesWithUnpromotableShardsStarted` is non null then we should always have removed a block?

Reply:
It depends, in cases where the index has multiple shards it might be possible that only one unpromotable shard has started and therefore we shouldn't remove the block.

Reply:
Oh yes of course.